File: leetcode/298-binary-tree-longest-consecutive-sequence.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
 public:
  int longestConsecutive(TreeNode* root)
  {
    maxlen_ = 0;
    if (root) dfs(root, 1);
    return maxlen_;
  }

  void dfs(TreeNode* node, int len)
  {
    if (len > maxlen_) maxlen_ = len;
    for (TreeNode* child : {node->left, node->right}) {
      if (!child) continue;
      int cur = node->val - child->val;
      if (-1 == cur) dfs(child, len+1);
      else dfs(child, 1);
    }
  }

 private:
  int maxlen_;
};

File: leetcode/298-binary-tree-longest-consecutive-sequence.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
  def longestConsecutive(self, root):
    """
    :type root: TreeNode
    :rtype: int
    """
    def _dfs(node, maxlen):
      l, r = 1, 1
      if node.left:
        dl = _dfs(node.left, maxlen)
        if node.val+1 == node.left.val:
          l += dl
      if node.right:
        dr = _dfs(node.right, maxlen)
        if node.val+1 == node.right.val:
          r += dr

      if l > r:
        if l > maxlen['val']:
          maxlen['val'] = l
        return l

      if r > maxlen['val']:
        maxlen['val'] = r
      return r

    maxlen = {'val': 0}
    if root:
      _dfs(root, maxlen)
    return maxlen['val']
